src/main/java/org/jenkinsci/remoting/util/VersionNumber.java

```diff
@@ -0,0 +1,579 @@
+                }
+                startIndex = i + 1;
+            } else if (c == '-') {
+                if (i == startIndex) {
+                    list.add(IntegerItem.ZERO);
+                } else {
+                    list.add(parseItem(isDigit, version.substring(startIndex, i)));
+                }
+                startIndex = i + 1;
+
+                if (isDigit) {
+                    list.normalize(); // 1.0-* = 1-*
+
+                    if ((i + 1 < version.length()) && Character.isDigit(version.charAt(i + 1))) {
+                        // new ListItem only if previous were digits and new char is a digit,
+                        // ie need to differentiate only 1.1 from 1-1
+                        list.add(list = new ListItem());
+
+                        stack.push(list);
+                    }
+                }
+            } else if (c == '*') {
+                list.add(new WildCardItem());
+                startIndex = i + 1;
+            } else if (Character.isDigit(c)) {
+                if (!isDigit && i > startIndex) {
+                    list.add(new StringItem(version.substring(startIndex, i), true));
+                    startIndex = i;
+                }
+
+                isDigit = true;
+            } else if (Character.isWhitespace(c)) {
+                if (i > startIndex) {
+                    if (isDigit) {
+                        list.add(parseItem(true, version.substring(startIndex, i)));
+                    } else {
+                        list.add(new StringItem(version.substring(startIndex, i), true));
+                    }
+                    startIndex = i;
+                }
+
+                isDigit = false;
+            } else {
+                if (isDigit && i > startIndex) {
+                    list.add(parseItem(true, version.substring(startIndex, i)));
+                    startIndex = i;
+                }
+
+                isDigit = false;
+            }
+        }
+
+        if (version.length() > startIndex) {
+            list.add(parseItem(isDigit, version.substring(startIndex)));
+        }
+
+        while (!stack.isEmpty()) {
+            list = (ListItem) stack.pop();
+            list.normalize();
+        }
+
+        canonical = items.toString();
+    }
+
+    private static Item parseItem(boolean isDigit, String buf) {
+        return isDigit ? (Item) new IntegerItem(buf) : (Item) new StringItem(buf, false);
+    }
+
+    public int compareTo(VersionNumber o) {
+        int result = items.compare(o.items);
+        if (result != 0) {
+            return result;
+        }
+        if (snapshot == null) {
+            return o.snapshot == null ? 0 : -1;
+        }
+        if (o.snapshot == null) {
+            return 1;
+        }
+        if ("-SNAPSHOT".equals(snapshot) || "-SNAPSHOT".equals(o.snapshot)) {
+            // cannot compare literal with timestamped.
+            return 0;
```

Comment:
So they're considered equal then? Maybe this should throw an `IllegalArgumentException` instead.